rasa/core/validator.py

```diff
@@ -32,124 +32,142 @@ def __init__(self, domain: Domain, intents: TrainingData, stories: List[StorySte
 
         for action in actions:
             if action.startswith(UTTER_PREFIX):
                 if action not in utterance_templates:
                     logger.error(
                         "There is no template for utterance '{}'.".format(action)
                     )
+                    everything_is_alright = False
 
-        return valid_utterances
+        return everything_is_alright
 
-    def verify_utterances_in_stories(self):
+    def verify_utterances_in_stories(self, ignore_warnings: bool = True) -> bool:
         """Verifies usage of utterances in stories.
 
         Checks whether utterances used in the stories are valid,
         and whether all valid utterances are used in stories."""
 
-        valid_utterances = self.verify_utterances()
+        everything_is_alright = self.verify_utterances()
+
+        utterance_actions = self._gather_utterance_actions()
         stories_utterances = set()
 
         for story in self.stories:
             for event in story.events:
-                if isinstance(event, ActionExecuted) and event.action_name.startswith(
-                    UTTER_PREFIX
-                ):
-                    utterance = event.action_name
-                    if (
-                        utterance not in valid_utterances
-                        and utterance not in stories_utterances
-                    ):
-                        logger.error(
-                            "The utterance '{}' is used in stories, but is not a "
-                            "valid utterance.".format(utterance)
-                        )
-                    stories_utterances.add(utterance)
-
-        for utterance in valid_utterances:
+                if not isinstance(event, ActionExecuted):
+                    continue
+                if not event.action_name.startswith(UTTER_PREFIX):
+                    # we are only interested in utter actions
+                    continue
+
+                if event.action_name in stories_utterances:
+                    # we already processed this one before, we only want to warn once
+                    continue
+
+                if event.action_name not in utterance_actions:
+                    logger.error(
+                        "The utterance '{}' is used in stories, but is not a "
+                        "valid utterance.".format(event.action_name)
+                    )
+                    everything_is_alright = False
+                stories_utterances.add(event.action_name)
+
+        for utterance in utterance_actions:
             if utterance not in stories_utterances:
                 logger.warning(
                     "The utterance '{}' is not used in any story.".format(utterance)
                 )
+                everything_is_alright = ignore_warnings and everything_is_alright
+
+        return everything_is_alright
 
-    def verify_all(self):
+    def verify_all(self, ignore_warnings: bool = True) -> bool:
```

Comment:
does it make more sense just to pass a `fail_on_warnings` argument instead of weirdly having to pass `not args.fail_on_warnings` to the function?

Reply:
I had this before, but that makes writing the statements
` everything_is_alright = ignore_warnings and everything_is_alright` a lot harder. so I went for the single `not` instead of having it in each of these statements.

Reply:
makes sense!